File: src/main/java/org/fountainmc/api/BlockType.java
package org.fountainmc.api;

import org.fountainmc.api.world.block.BlockState;

public interface BlockType extends Material {

    @Override
    default boolean isBlock() {
        return true;
    }

    boolean isFlammable();

    /**
     * Return if the block completely blocks all vision
     *
     * @return if opaque
     */
    boolean isOpaque();

    /**
     * Return if the block is transparent and the player can see through it
     * <p>Transparent blocks do not block light.</p>
     *
     * @return if transparent
     */
    boolean isTransparent();

    /**
     * Get the default state the block is in
     *
     * @return the default state of the block
     */
    BlockState getDefaultState();

}

File: src/main/kotlin/org/fountainmc/api/BlockType.kt
package org.fountainmc.api

import org.fountainmc.api.inventory.item.BlockItem
import org.fountainmc.api.world.block.BlockState
import kotlin.reflect.KClass

interface BlockType<out S: BlockState> : Material<BlockItem<S>> {

    val stateType: KClass<@UnsafeVariance S>

    override val isBlock
        get() = true

    /**
     * If the block is flammable.
     */
    val isFlammable: Boolean

    /**
     * If the block completely blocks all vision
     */
    val isOpaque: Boolean

    /**
     * If the block is transparent and the player can see through it
     *
     * Transparent blocks do not block light.
     */
    val isTransparent: Boolean

    /**
     * The default state the block is in.
     */
    val defaultState: S
}
